File: java/350intersection-of-two-arrays-ii/350.两个数组的交集-ii.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

/*
 * @lc app=leetcode.cn id=350 lang=java
 *
 * [350] 两个数组的交集 II
 */

// @lc code=start
class Solution {
    /**
     * 1、暴力法：两层循环，时间复杂度 k*n
     * 2、其中较小的数组放到 hashMap 内，key是值，value是出现次数。时间复杂度 k+n
     * 3、先将两个数组排序，利用双指针在一次循环内解决
     * @param nums1
     * @param nums2
     * @return
     */
    public int[] intersect(int[] nums1, int[] nums2) {
        if(nums1.length > nums2.length) {
            return intersect(nums2, nums1);
        }
        Map<Integer, Integer> num1Map = new HashMap<>();
        for (int n1 : nums1) {
            Integer count = num1Map.get(n1);
            if (count == null) {
                count = 1;
            } else {
                count++;
            }
            num1Map.put(n1, count);
        }
        int[] result = nums1;
        int resultCount = 0;
        for (int n2 : nums2) {
            Integer count = num1Map.get(n2);
            if (count != null && count > 0) {
                result[resultCount] = n2;
                resultCount++;
                count--;
                num1Map.put(n2, count);
            }
        }
        return Arrays.copyOfRange(result, 0, resultCount);
    }
}
// @lc code=end

File: go/350intersect/350.两个数组的交集-ii.go
/*
 * @lc app=leetcode.cn id=350 lang=golang
 *
 * [350] 两个数组的交集 II
 */

// @lc code=start
func intersect(nums1 []int, nums2 []int) []int {
	l1, l2 := len(nums1), len(nums2)
	if l1 == 0 || l2 == 0 {
		return []int{}
	}
	result := make([]int, 0)
	if l1 > l2 {
		l1, l2 = l2, l1
		nums1, nums2 = nums2, nums1
	}
	nums2Map := make(map[int]int, l2)
	for _, v := range nums2 {
		count := nums2Map[v]
		count++
		nums2Map[v] = count
	}
	for _, v := range nums1 {
		if count, ok := nums2Map[v]; ok {
			result = append(result, v)
			if count > 1 {
				count--
				nums2Map[v] = count
			} else {
				delete(nums2Map, v)
			}
		}
	}
	return result
}
// @lc code=end

